lib/Conversion/FSMToSV/FSMToSV.cpp

```diff
@@ -0,0 +1,564 @@
+
+public:
+  StateEncoding(OpBuilder &b, MachineOp machine, hw::HWModuleOp hwModule)
+      : b(b), machine(machine), hwModule(hwModule) {}
+  virtual ~StateEncoding() {}
+
+  Value encode(StateOp state) {
+    auto it = stateToValue.find(state);
+    assert(it != stateToValue.end() && "state not found");
+    return it->second;
+  }
+  StateOp decode(Value value) {
+    auto it = valueToState.find(value);
+    assert(it != valueToState.end() && "encoded state not found");
+    return it->second;
+  }
+
+  // Returns the type which encodes the state values.
+  virtual Type getStateType() = 0;
+
+  // Returns a case pattern which matches the provided state.
+  virtual std::unique_ptr<sv::CasePattern> getCasePattern(StateOp state) = 0;
+
+protected:
+  // Creates a constant value in the module for the given encoded state
+  // and records the state value in the mappings. An inner symbol is
+  // attached to the wire to avoid it being optimized away.
+  // The constant can optionally be assigned behind a sv wire - doing so at this
+  // point ensures that constants don't end up behind "_GEN#" wires in the
+  // module.
+  void setEncoding(StateOp state, Value v, bool wire = false) {
+    assert(stateToValue.find(state) == stateToValue.end() &&
+           "state already encoded");
+
+    Value encodedValue;
+    if (wire) {
+      auto loc = machine.getLoc();
+      auto stateType = getStateType();
+      auto stateEncodingWire = b.create<sv::WireOp>(
+          loc, stateType, b.getStringAttr("to_" + state.getName()),
+          /*inner_sym=*/state.getNameAttr());
+      b.create<sv::AssignOp>(loc, stateEncodingWire, v);
+      encodedValue = b.create<sv::ReadInOutOp>(loc, stateEncodingWire);
+    } else
+      encodedValue = v;
+    stateToValue[state] = encodedValue;
+    valueToState[encodedValue] = state;
+    valueToSrcValue[encodedValue] = v;
+  }
+
+  // A mapping between a StateOp and its corresponding encoded value.
+  SmallDenseMap<StateOp, Value> stateToValue;
+
+  // A mapping between an encoded value and its corresponding StateOp.
+  SmallDenseMap<Value, StateOp> valueToState;
+
+  // A mapping between an encoded value and the source value in the IR.
+  SmallDenseMap<Value, Value> valueToSrcValue;
+
+  OpBuilder &b;
+  MachineOp machine;
+  hw::HWModuleOp hwModule;
+};
+
+class EnumStateEncoding : public StateEncoding {
+public:
+  EnumStateEncoding(OpBuilder &b, MachineOp machine, hw::HWModuleOp hwModule)
+      : StateEncoding(b, machine, hwModule) {
+
+    Location loc = machine.getLoc();
+    llvm::SmallVector<Attribute> stateNames;
+
+    for (auto state : machine.getBody().getOps<StateOp>())
+      stateNames.push_back(b.getStringAttr(state.getName()));
+
+    // Create an enum typedef for the states.
+    Type rawEnumType =
+        hw::EnumType::get(b.getContext(), b.getArrayAttr(stateNames));
+
+    OpBuilder::InsertionGuard guard(b);
+    b.setInsertionPoint(hwModule);
+    auto typeScope = b.create<hw::TypeScopeOp>(
```

Comment:
@mikeurbach Would it make sense to have a typedef pass to create typescopes for every typealias in the IR being emitted? We have to do the same thing in PyCDE IIRC. I seem to recall a reason we don't have this pass already...